soroban-env-host/src/host/prng.rs

```diff
@@ -0,0 +1,151 @@
+///     anything they like with, including copying to guest linear memory), and
+///     two slightly more subtle but very standard operations that are easy to
+///     get wrong: an inclusive-range uniform u64 sampler and a Fisher-Yates
+///     vector shuffle. The latter is also hard to do cheaply in guest code
+///     without copying the vector into the guest and copying it back.
+///
+///   - All these PRNGs are ChaCha20: a strong, cheap, standard CSPRNG.
+///
+#[derive(Debug, Clone)]
+pub(crate) struct Prng(ChaCha20Rng);
+
+pub type Seed = <rand_chacha::ChaCha20Rng as rand::SeedableRng>::Seed;
+pub const SEED_BYTES: usize = core::mem::size_of::<Seed>();
+static_assertions::const_assert_eq!(SEED_BYTES, 32);
+
+impl Default for Prng {
+    fn default() -> Self {
+        Self(ChaCha20Rng::from_seed([0; SEED_BYTES]))
+    }
+}
+
+impl Prng {
+    fn charge_prng_bytes(&self, budget: &Budget, count: u64) -> Result<(), HostError> {
+        // TODO: add a ContractCostType for drawing PRNG bytes
+        Ok(())
+    }
+
+    pub(crate) fn reseed(&mut self, seed: Seed) {
+        self.0 = ChaCha20Rng::from_seed(seed);
+    }
+
+    pub(crate) fn u64_in_inclusive_range(
+        &mut self,
+        range: RangeInclusive<u64>,
+        budget: &Budget,
+    ) -> Result<u64, HostError> {
+        // We over-estimate the number of bytes drawn by a factor of 2, to
+        // account for the fact that a range sample is rejection-sampling which
+        // is expected to only do one draw but might do more than one.
+        self.charge_prng_bytes(budget, (2 * core::mem::size_of::<u64>()) as u64)?;
+        let u = Uniform::from(range);
+        Ok(u.sample(&mut self.0))
+    }
+
+    pub(crate) fn vec_shuffle(
+        &mut self,
+        v: &HostVec,
+        budget: &Budget,
+    ) -> Result<HostVec, HostError> {
+        // A Fisher-Yates shuffle essentially does one call to u64_in_range for
+        // each element of the input vector, followed by an optional swap. Since
+        // u64_in_range is itself a rejection sampling operation (to avoid bias)
+        // we can't be 100% sure how many draws it'll make, but the expected
+        // number of draws is 1. To give ourselves a little more safety we'll
+        // double that number. We also give the implementation freedom to draw a
+        // 64-bit (8-byte) value per index, meaning we charge for generating 2 *
+        // 8 * len bytes.
+        let mut v2 = v.metered_clone(budget)?;
+        // We charge for both the PRNG draws and the swaps here (as "memcpys").
+        self.charge_prng_bytes(budget, 16u64.saturating_mul(v.len() as u64))?;
+        budget.charge(ContractCostType::HostMemCpy, Some(v.len() as u64))?;
+        v2.as_mut_slice().shuffle(&mut self.0);
+        Ok(v2)
+    }
+
+    pub(crate) fn bytes_new(&mut self, size: u32, budget: &Budget) -> Result<ScBytes, HostError> {
+        budget.charge(ContractCostType::HostMemAlloc, Some(size as u64))?;
```

Comment:
passing thought: it'd be really cool if the **only** way to get host memory was via charging, rather than relying on ourselves to ensure we always charge the budget when appropriate. but maybe that's too restrictive? dunno, just a drive-by review musing